scipy/stats/_qmc_cy.pyx

```diff
@@ -326,3 +326,43 @@ cdef void one_thread_loop(func_type loop_func, double& disc, double[:,
     threaded_sum_mutex.lock()
     (&disc)[0] += tmp # workaround to "disc += tmp", see cython issue #1863
     threaded_sum_mutex.unlock()
+
+
+def _cy_van_der_corput(long n, long base, long start_index):
+    sequence = np.zeros(n)
+    cdef:
+        long i, quotient, remainder
+        double[::1] sequence_view = sequence
+        double b2r
+
+    for i in range(n):
+        quotient = start_index + i
+        b2r = 1.0 / base
+        while quotient > 0:
```

Comment:
There are fewer iterations in this `while` loop than in the `for` loop in `_cy_van_der_corput_scrambled`. In the unscrambled case, there is a finite number of iterations needed to generate each element in the Van der Corput sequence. Any further iterations would only add 0 to each element.  For the scrambled case, you keep iterating until the addition to the element is smaller than precision of the float.

Reply:
I added a benchmark for Halton sequences to this branch. I also made a new branch that uses a while loop instead of a for loop.  That new branch is here https://github.com/treverhines/scipy/tree/cythonize-van-der-corput-with-while-loop

The difference between the two branches (from `cythonize-van-der-corput` to `cythonize-van-der-corput-with-while-loop`) is this:
```diff
diff --git a/benchmarks/asv.conf.json b/benchmarks/asv.conf.json
index 5eec9cf51..ca9f1e193 100644
--- a/benchmarks/asv.conf.json
+++ b/benchmarks/asv.conf.json
@@ -40,7 +40,7 @@
         "numpy": [],
         "Cython": ["0.29.21"],
         "pytest": [],
-        "pythran": [],
+        "pip+pythran": [],
         "pybind11": [],
     },

diff --git a/scipy/stats/_qmc_cy.pyx b/scipy/stats/_qmc_cy.pyx
index 6d72b3b57..22ac082b0 100644
--- a/scipy/stats/_qmc_cy.pyx
+++ b/scipy/stats/_qmc_cy.pyx
@@ -358,11 +358,13 @@ def _cy_van_der_corput_scrambled(long n, long base, long start_index,
     for i in range(n):
         quotient = start_index + i
         b2r = 1.0 / base
-        for j in range(permutations.shape[0]):
+        j = 0
+        while (1 - b2r) < 1:
             remainder = quotient % base
             remainder = permutations[j, remainder]
             sequence_view[i] += remainder * b2r
             b2r /= base
             quotient //= base
+            j += 1

     return sequence
```
I needed to modify `asv.conf.json` so that ASV would use pip to install pythran for the benchmarking environment. Otherwise, it would check for pythran in the default conda channel and it would fail. 

While I was on branch `cythonize-van-der-corput-with-while-loop`, I compared the Halton benchmark on the two branches using the command `python runtests.py --bench-compare cythonize-van-der-corput --bench BenchQMCHalton`. The benchmark comparison finished after about 10 minutes and this is the output:
```
· No `environment_type` specified in asv.conf.json. This will be required in the future.
· Creating environments
· Discovering benchmarks.
·· Uninstalling from conda-py3.8-Cython0.29.21-numpy-pip+pythran-pybind11-pytest
·· Building 38ac5f88 <cythonize-van-der-corput-with-while-loop> for conda-py3.8-Cython0.29.21-numpy-pip+pythran-pybind11-pytest................................................................................................................................................................
·· Installing 38ac5f88 <cythonize-van-der-corput-with-while-loop> into conda-py3.8-Cython0.29.21-numpy-pip+pythran-pybind11-pytest.
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scipy commit 5de221a6 <cythonize-van-der-corput> (round 1/2):
[  0.00%] ·· Building for conda-py3.8-Cython0.29.21-numpy-pip+pythran-pybind11-pytest..
[  0.00%] ·· Benchmarking conda-py3.8-Cython0.29.21-numpy-pip+pythran-pybind11-pytest
[ 25.00%] ··· Running (stats.BenchQMCHalton.time_discrepancy--).
[ 25.00%] · For scipy commit 38ac5f88 <cythonize-van-der-corput-with-while-loop> (round 1/2):
[ 25.00%] ·· Building for conda-py3.8-Cython0.29.21-numpy-pip+pythran-pybind11-pytest..
[ 25.00%] ·· Benchmarking conda-py3.8-Cython0.29.21-numpy-pip+pythran-pybind11-pytest
[ 50.00%] ··· Running (stats.BenchQMCHalton.time_discrepancy--).
[ 50.00%] · For scipy commit 38ac5f88 <cythonize-van-der-corput-with-while-loop> (round 2/2):
[ 50.00%] ·· Benchmarking conda-py3.8-Cython0.29.21-numpy-pip+pythran-pybind11-pytest
[ 75.00%] ··· stats.BenchQMCHalton.time_discrepancy                                                                          ok
[ 75.00%] ··· ==== ============= ============= =============== ============ ============== ================
              --                                         scramble / n                                      
              ---- ----------------------------------------------------------------------------------------
               d     True / 10    True / 1000   True / 100000   False / 10   False / 1000   False / 100000 
              ==== ============= ============= =============== ============ ============== ================
               1     216±0.8μs      722±2μs      51.8±0.09ms    115±0.9μs     203±0.3μs      16.2±0.06ms   
               10   1.09±0.01ms   3.03±0.02ms     200±0.5ms      592±2μs       933±4μs        65.3±0.1ms   
              ==== ============= ============= =============== ============ ============== ================

[ 75.00%] · For scipy commit 5de221a6 <cythonize-van-der-corput> (round 2/2):
[ 75.00%] ·· Building for conda-py3.8-Cython0.29.21-numpy-pip+pythran-pybind11-pytest..
[ 75.00%] ·· Benchmarking conda-py3.8-Cython0.29.21-numpy-pip+pythran-pybind11-pytest
[100.00%] ··· stats.BenchQMCHalton.time_discrepancy                                                                          ok
[100.00%] ··· ==== =========== ============= =============== ============ ============== ================
              --                                        scramble / n                                     
              ---- --------------------------------------------------------------------------------------
               d    True / 10   True / 1000   True / 100000   False / 10   False / 1000   False / 100000 
              ==== =========== ============= =============== ============ ============== ================
               1     216±2μs      721±1μs      51.6±0.02ms    115±0.7μs     203±0.7μs      16.2±0.07ms   
               10    1.09±0ms   3.02±0.04ms     200±0.07ms     594±7μs      937±0.8μs      65.2±0.02ms   
              ==== =========== ============= =============== ============ ============== ================


BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

So I am not able to reproduce the timing difference that you are seeing. If you run this benchmark comparison, do you see something else?